src/Nethermind.Crypto.SecP256r1/DerHelper.cs

```diff
@@ -0,0 +1,52 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+namespace Nethermind.Crypto;
+
+internal static class DerHelper
+{
+    // Max possible size for DER-encoded array of the P256 signature:
+    // 2 for SEQUENCE header and length
+    // for each of 2 elements (R, S):
+    // - 2 for INTEGER header and length
+    // - 1 for the leading zero byte if number starts with "negative" byte
+    // - 32 for the element itself
+    public const int P256SignatureMaxSize = 2 + 2 * (2 + 1 + 32);
+
+    // Encodes signature (r,s) in DER format
+    // AsnWriter allocates too much
+    public static ReadOnlySpan<byte> EncodeP256Signature(ReadOnlySpan<byte> r, ReadOnlySpan<byte> s, Span<byte> buffer)
+    {
+        buffer[0] = 0x30; // SEQUENCE OF
+
+        var index = 2;
+        index += EncodeUnsignedInteger(r, buffer[index..]);
+        index += EncodeUnsignedInteger(s, buffer[index..]);
+
+        buffer[1] = (byte)(index - 2); // SEQUENCE OF length
+
+        return buffer[..index];
+    }
+
+    private static int EncodeUnsignedInteger(ReadOnlySpan<byte> value, Span<byte> buffer)
+    {
+        // Skip zeroes
+        var valIndex = 0;
+        while (value[valIndex] == 0 && valIndex < value.Length - 1) valIndex++;
```

Comment:
How often does it happen? How big is the input? Have you considered using vectorized `.IndexOfAnyExcept` or `BitOperations.LeadingZeroCount` if value is up to 8 bytes?

Reply:
`value` is always 32 bytes. Not sure about number-of-leading-zero-bytes distribution, I guess theoretically it can be any value from 0 to 31, with an average probably something between 2 and 6.

Will try check `IndexOfAnyExcept` is faster, but in general I don't expect this part to take anything even close to 1% of the total execution time.

Reply:
If average case is 2 to 6, then `BitOperations.LeadingZeroCount(Unsafe.ReadUnaligned<ulong>(ref value[0))` would give you the answer with a single instruction (or `BitOperations.TrailingZeroCount` for little endian). `IndexOfAnyExcept` won't be probably faster because at the end it performs the call. Your call :-)

Reply:
Never mind, seems like I managed to get rid of the DER-encoding phase as a whole.